### PR TITLE
Plugin input

### PR DESCRIPTION
**Pull request by brunnre8**
This types the public command API during the registration phase and reworks the input function to be a bit easier to understand, least for me.

Holler if you disagree, but I hate the pattern of

```js
if(stuff) {
 // ...
} else if (other) {
 // ...
} else if (yet another) {
 // ...
}
// return
```

That (implicit) return right there is not even on screen when you start to read the first few clauses, so you have no idea what the function does at the end only to then scroll down and find out that it indeed just returns.

We already know that we are done, tell the reader accordingly.

### CODE REVIEW

File: server/client.ts [hunk 1]
```diff
@@ -465,46 +465,54 @@ class Client {
...
+				new Msg({
+					type: MessageType.ERROR,
+					text: "You are not connected to the IRC network, unable to send your command.",
+				})
+			);
+		};
 
 		if (inputs.userInputs.has(cmd)) {
-			const plugin = inputs.userInputs.get(cmd);
+			const plugin = inputs.userInputs.get(cmd)!; // this is only save because we just checked it
```
Comment (MaxLeiter) [inline server/client.ts:481]: nit: I meant to suggest changes here, but I'd prefer something like https://github.com/thelounge/thelounge/pull/4714

File: server/plugins/inputs/index.ts [hunk 2]
```diff
@@ -89,9 +101,22 @@ const getCommands = () =>
 		.concat(passThroughCommands)
 		.sort();
 
-const addPluginCommand = (packageInfo: PackageInfo, command, func) => {
-	func.packageInfo = packageInfo;
-	pluginCommands.set(command, func);
+const addPluginCommand = (packageInfo: PackageInfo, command: any, obj: any) => {
```
Comment (MaxLeiter) [inline server/plugins/inputs/index.ts:104]: should `obj: any` be `obj: ExternalPluginCommand`?

Comment (brunnre8) [reply server/plugins/inputs/index.ts:104]: nope, it's provided by an external package, we literally don't know what it is, we must kot lie to the compiler / human

File: server/client.ts (hunk 1 above)
Comment (brunnre8) [reply server/client.ts:481]: your PR over there fails tze typing check though

File: server/plugins/inputs/index.ts (hunk 2 above)
Comment (MaxLeiter) [reply server/plugins/inputs/index.ts:104]: nit: `unknown` is preferred then

**Review by MaxLeiter — APPROVED** [review]

File: server/plugins/inputs/index.ts (hunk 2 above)
Comment (brunnre8) [reply server/plugins/inputs/index.ts:104]: ack, will change

Comment (brunnre8) [reply server/plugins/inputs/index.ts:104]: actually... if I do that the typing gets really annoying.

It errors out even if I guard that this is an object, due to it not having the input property.
Now, I can either type assert it or just keep it as any.

Type asserting it seems worse to me, opinons?

File: server/client.ts (hunk 1 above)
Comment (brunnre8) [reply server/client.ts]: done

File: server/plugins/inputs/index.ts (hunk 2 above)
Comment (MaxLeiter) [reply server/plugins/inputs/index.ts:104]: Then I think its fine as is. Sorry for wasting your time

**Pull request merged**